File: day_1/day_1.py
from pathlib import Path


def parse_input():
    input_file = Path(__file__).parent / "input.txt"
    return [int(line.strip()) for line in input_file.open()]


def part1(data: list[int]) -> int:
    tmp: int = data[0]
    counter: int = 0
    for depth in data[1:]:
        if depth > tmp:
            counter += 1
        tmp = depth
    return counter


def part2(data: list[int]):
    window_a: int = sum(data[:3])
    counter: int = 0
    for i in range(1, len(data) - 2):
        window_b = sum(data[i : i + 3])
        if window_b > window_a:
            counter += 1
        window_a = window_b
    return counter


def main():
    data = parse_input()
    print(f"{part1(data)=}")
    print(f"{part2(data)=}")


if __name__ == "__main__":
    main()

File: day_1/day_1.nim
import std/[strutils, strformat, sugar]
from std/os import parentDir

proc parse_input(): seq[int] =
  let file = currentSourcePath.parentDir() & "\\input.txt"
  result = collect(newSeq):
    for line in file.lines: line.strip.parseInt

proc part1(depths: seq[int]): int =
  var
    tmp: int = depths[0]
  for depth in depths:
    if depth > tmp:
      result += 1
    tmp = depth

proc part2(depths: seq[int]): int =
  for i, depth in depths[0..^4]:
    if depth < depths[i+3]:
      result += 1

proc main(): void =
  let data = parse_input()
  echo(fmt"{part1(data)=}")
  echo(fmt"{part2(data)=}")

when is_main_module:
  main()
